src/libexpr/eval.cc

```diff
@@ -65,26 +65,18 @@ static char * dupString(const char * s)
 
 // When there's no need to write to the string, we can optimize away empty
 // string allocations.
-// This function handles makeImmutableStringWithLen(null, 0) by returning the
-// empty string.
-static const char * makeImmutableStringWithLen(const char * s, size_t size)
-{
-    char * t;
+// This function handles makeImmutableString(std::string_view()) by returning
+// the empty string.
+static inline const char * makeImmutableString(std::string_view s) {
+    size_t size = s.size();
     if (size == 0)
         return "";
-#if HAVE_BOEHMGC
-    t = GC_STRNDUP(s, size);
-#else
-    t = strndup(s, size);
-#endif
-    if (!t) throw std::bad_alloc();
+    char * t = allocString(size + 1);
+    memcpy(t, s.data(), size);
```

Comment:
`strncpy` could be used instead and an additional check can be added, which ensures that the string_view did contain no zeroes.
The drawback might be reduced performance

Reply:
I worked around this [a while ago](https://github.com/NixOS/nix/pull/7249/commits/742654c5b09f824167b374cc66647e5d0abb1df1) using `strnlen`

Reply:
@yorickvP `strnlen` and `memcpy` is still slower than a single `memcpy` and probably even slower than `strncpy` because in the normal case `strnlen` would still scan the whole string.